### ISSUE TITLE
panic: send on closed channel

### ISSUE THREAD

**Issue by agupta7** [squat/generic-device-plugin#37]
There's a panic happening when one of the latter two functions in the run group return.

```
{"caller":"main.go:218","msg":"Starting the generic-device-plugin for \"squat.ai/device\".","ts":"2023-08-16T21:15:42.090254018Z"}
{"caller":"plugin.go:114","level":"info","msg":"listening on Unix socket","resource":"squat.ai/device","socket":"/var/lib/kubelet/device-plugins/gdp-c3F1YXQuYWkvd2F0Y2hkb2c=-1692220542.sock","ts":"2023-08-16T21:15:42.090439622Z"}
{"caller":"plugin.go:122","level":"info","msg":"starting gRPC server","resource":"squat.ai/device","ts":"2023-08-16T21:15:42.090637276Z"}
{"caller":"plugin.go:174","level":"info","msg":"waiting for the gRPC server to be ready","resource":"squat.ai/device","ts":"2023-08-16T21:15:42.090677679Z"}
{"caller":"plugin.go:186","level":"info","msg":"the gRPC server is ready","resource":"squat.ai/device","ts":"2023-08-16T21:15:42.091178516Z"}
{"caller":"plugin.go:224","level":"info","msg":"registering plugin with kubelet","resource":"squat.ai/device","ts":"2023-08-16T21:15:42.091192467Z"}
{"caller":"generic.go:225","level":"info","msg":"starting listwatch","resource":"squat.ai/device","ts":"2023-08-16T21:15:42.187816593Z"}
panic: send on closed channel

goroutine 22 [running]:
github.com/squat/generic-device-plugin/deviceplugin.(*plugin).serve.func1()
        /src/deviceplugin/plugin.go:123 +0x12b
created by github.com/squat/generic-device-plugin/deviceplugin.(*plugin).serve
        /src/deviceplugin/plugin.go:121 +0x378
```

This panic masks the logging of the actual error returned by the returning function in the run group and prevents the device-plugin from retrying.  I've seen hundreds of restarts from such panics over a few days.

**Comment by squat** [comment]
Thanks for reporting this issue, @agupta7! I just posted a fix in #41. Please try the latest image of the device plugin once the branch is merged <3

**Comment by aledbf** [comment]
@squat I see a similar issue (no panic but a CrashLoopBackOff) when I delete a pod with the annotations
The latest line I see:
```
{"caller":"generic.go:232","level":"info","msg":"starting listwatch","resource":"squat.ai/my-device","ts":"2023-09-16T21:00:40.543869325Z"}
```

**Comment by squat** [comment]
Interesting! This is using the latest image? Can you share your configuration/info about your cluster so I can try to reproduce? Let's open a new issue!

**Comment by aledbf** [comment]
> This is using the latest image? 

Yes.

I think this is related to the resource limits:
```
# kubectl top pods -n kube-system
NAME                                       CPU(cores)   MEMORY(bytes)   
calico-kube-controllers-85c96b4ffc-pwd7s   1m           58Mi            
calico-node-wkgmg                          17m          122Mi           
coredns-77ccd57875-mmhvw                   1m           57Mi            
generic-device-plugin-nbs92                3m           10Mi            
local-path-provisioner-957fdf8bc-45c9p     1m           33Mi            
metrics-server-648b5df564-kbwzl            2m           61Mi  
```

After increasing the memory limit to 100Mi I don't see the issue.

> Server Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.5+k3s1", GitCommit:"8d074ecb5a8765a09eeef6f8be7987055210bc40", GitTreeState:"clean", BuildDate:"2023-09-05T18:24:36Z", GoVersion:"go1.20.7", Compiler:"gc", Platform:"linux/amd64"}
